File: src/main/java/org/academy/internal/client/app/music/backend/engine/AudioPlayer.java
package org.academy.internal.client.app.music.backend.engine;

import org.academy.AcademyCraft;
import org.jspecify.annotations.Nullable;
import org.slf4j.Logger;

import java.io.IOException;
import java.nio.ByteBuffer;

public final class AudioPlayer {
    private static final Logger LOGGER = AcademyCraft.getLogger();

    @Nullable
    private StreamingPlaybackSession currentSession;
    private float volume = 1.0f;

    public void play(ByteBuffer audioData, long startFrame) {
        stop();
        try {
            currentSession = new StreamingPlaybackSession(audioData, startFrame, volume);
        } catch (IOException e) {
            LOGGER.error("Failed to start audio playback session", e);
            stop();
        }
    }

    public void stop() {
        if (currentSession != null) {
            currentSession.destroy();
            currentSession = null;
        }
    }

    public void update() {
        if (currentSession != null) currentSession.update();
    }

    public void togglePlayPause() {
        if (currentSession != null) currentSession.togglePlayPause();
    }

    public void seek(ByteBuffer audioData, float timeRatio) {
        if (currentSession != null) {
            var totalSamples = currentSession.getTotalSamples();
            var targetFrame = (long) (timeRatio * totalSamples);
            currentSession.seek(targetFrame);
        } else play(audioData, 0);
    }

    public void handleContextReset() {
        LOGGER.info("Sound engine reloaded, resetting media player OpenAL resources.");
        stop();
    }

    public void setVolume(float value) {
        volume = value;
        if (currentSession != null) currentSession.setVolume(value);
    }

    public float getCurrentTime() {
        return currentSession != null ? currentSession.getCurrentTime() : 0;
    }

    public boolean isPlaying() {
        return currentSession != null && currentSession.isPlaying();
    }

    public boolean isPaused() {
        return currentSession != null && currentSession.isPaused();
    }

    public boolean isFinished() {
        return currentSession == null || currentSession.isFinished();
    }

    public float getTotalDuration() {
        return currentSession != null ? currentSession.getTotalDuration() : 0f;
    }

    public float getVolume() {
        return volume;
    }
}

File: src/main/kotlin/org/academy/internal/client/app/music/engine/AudioPlayer.kt
package org.academy.internal.client.app.music.engine

import org.academy.internal.client.app.music.common.PlaybackState
import java.nio.ByteBuffer
import java.util.concurrent.atomic.AtomicReference

class AudioPlayer {
    private var session: StreamingPlaybackSession? = null
    private val desiredState = AtomicReference(PlaybackState.IDLE)
    var volume: Float = 1.0f
        set(value) {
            field = value
            session?.setVolume(value)
        }

    val state: PlaybackState get() = desiredState.get()
    val currentTime: Float get() = session?.currentTime ?: 0f
    val totalDuration: Float get() = session?.totalDuration ?: 0f
    val isFinished: Boolean get() = session?.isFinished ?: true
    val hasSession: Boolean get() = session != null

    fun play(audioData: ByteBuffer, startFrame: Long) {
        stop()
        desiredState.set(PlaybackState.PLAYING)
        createSession(audioData, startFrame)
    }

    fun pause() {
        if (desiredState.get() == PlaybackState.PLAYING) desiredState.set(PlaybackState.PAUSED)
    }

    fun resume() {
        if (desiredState.get() == PlaybackState.PAUSED || (desiredState.get() == PlaybackState.IDLE && hasSession)) {
            desiredState.set(PlaybackState.PLAYING)
        }
    }

    fun togglePlayPause() {
        when (desiredState.get()) {
            PlaybackState.PLAYING -> pause()
            PlaybackState.PAUSED -> resume()
            PlaybackState.IDLE -> if (hasSession) resume()
        }
    }

    fun stop() {
        desiredState.set(PlaybackState.IDLE)
        session?.destroy()
        session = null
    }

    fun seek(audioData: ByteBuffer, timeRatio: Float) {
        val clampedRatio = timeRatio.coerceIn(0f, 1f)
        val session = session
        if (session != null) {
            val totalSamples = session.totalSamples
            val targetFrame = (clampedRatio * totalSamples).toLong()
            session.seek(targetFrame)
        } else {
            stop()
            desiredState.set(PlaybackState.IDLE)
            val newSession = createSession(audioData, 0L)
            val totalSamples = newSession.totalSamples
            val targetFrame = (clampedRatio * totalSamples).toLong()
            newSession.seek(targetFrame)
        }
    }

    fun update() {
        session?.let {
            it.update()
            if (it.isFinished) stop()
        }
    }

    private fun createSession(audioData: ByteBuffer, startFrame: Long): StreamingPlaybackSession {
        val newSession = StreamingPlaybackSession.create(audioData, startFrame, desiredState, volume)
        session = newSession
        return newSession
    }
}